### ISSUE TITLE
Impact of implicit stateful PyTrees?

### ISSUE THREAD

**Issue by nalzok** [patrick-kidger/equinox#93]
I find the following warning in the document

> `equinox.experimental.BatchNorm` updates its running statistics as a side effect of its forward pass. Side effects are quite unusual in JAX; as such `BatchNorm` is considered experimental. Let us know how you find it!

Could you elaborate on what could go wrong with such a side effect? I think it might be [the same catch](https://cgarciae.github.io/treex/user-guide/state-management/#what-is-the-catch) documented in treex, but I'm not sure if we need to output the Module to update its state in equinox since Python functions are kinda pass-by-reference, so the update should be reflected in the passed-in Module object without returning anything.

By the way, elegy/treex is planning on [implementing immutable PyTrees](https://github.com/cgarciae/treex/issues/69). Looks like you are taking different design paths and evolving separately!

**Comment by patrick-kidger** [comment]
**Re: comparisons to Treex**

It's not the same as in Treex. (This is definitely one place we diverge.)

In Equinox, then (a) `Module` is immutable, and (b) you don't need to pass a copy of a model out of a JIT'd region in order for its updates to be seen. In contrast Treex has mutable modules, and needs you to return copies of models in order to see updates. So I think the Equinox approach represents the best of both worlds here!

In terms of how Equinox manages this: loosely speaking, it works something like the following code.
```python
stats = {}

class BatchNorm(eqx.Module):
    index: object

    def __init__(self):
        self.index = object()

    def __call__(self, x):
        old_stat = stats[self.index]
        new_stat = jnp.some_statistic(x)
        stat = old_stat * 0.9 + new_stat * 0.1
        stats[self.index] = stat
        return x - stat
```
along with a bunch of stuff to make the updates work under JIT etc.

So far, I think Equinox is doing pretty well in this comparison! So it's worth being honest that there is a trade-off: Equinox's approach is probably going to be a bit slower on the GPU.

The thing is that Equinox aims to support much more complicated use-cases than just neural networks -- for example Equinox is what sits behind the differential equation solvers of [Diffrax](https://github.com/patrick-kidger/diffrax) -- so the underlying approach is very different.

I wanted to select an approach that -- whilst it might be slower -- is also more general + more robust + easier to use. Do let me know if the speed differences are enough to be noticeable/problematic. Also note that `eqx.tree_inference` can be used to toggle inference mode on `BatchNorm`, which amongst other things should remove any speed penalty at inference. (But prevent updates to the statistics during training, of course.)

**Re: things that can go wrong with side-effects**

Well, for example, maybe you're using some third-party library that assumes your function is pure. Then you pass in your neural network with BatchNorm into that library. Then you might get unexpected results. For example imagine including BatchNorm in an ODE solver: your vector field would be changing during the solve!

Now the reality is that of course you explicitly signed up for this by using a stateful operation like BatchNorm -- the assumption from the get-go is that you want side-effects because you used a layer whose whole purpose is to have side effects! So it should never really be a surprise to a user that they get funny behaviour if they use their side-effecting-neural-network-layer in complicated not-typical-neural-network ways. And after all, exactly the same is true in PyTorch! It's just that these are a bit unusual in JAX, so it's worth not forgetting that fact.

**Comment by nalzok** [comment]
Thanks for the response. It's really amazing that you can update a Python object in the global scope within JIT'd code!
